### PR TITLE
[OSS] test: refactor testcontainers and add peering integ tests

### PR DESCRIPTION
**Pull request by DanStough**
### Description
This PR adds additional integration tests for peering in advance of the 1.14 GA release. It also expands the capabilities of the testcontainer framework to run some connect-enabled workloads.

The tests are combined into a single case to amortize the cost of setting up a cluster.

The abstractions for the testcontainer library are still imperfect, but I expect these will improve over time.

### Testing & Reproduction steps
* You can run these with `go test` like ordinary unit tests. Docker must be installed and running.
* I would a review with `consul-containers/test/basic` to review the syntax and run.

### PR Checklist

* [X] updated test coverage
* [ ] ~ external facing docs updated~
* [ ] ~not a security concern~

### CODE REVIEW

File: test/integration/consul-container/libs/node/node.go
Comment (dhiaayachi) [inline test/integration/consul-container/libs/node/node.go:10]: At this point I think the name `Node` became confusing, maybe ConsulNode or even Consul?

**Review by dhiaayachi — APPROVED** [review]
This is awesome @DanStough!! 👏 
I especially like the assert package and the service interface.
I have small cosmetic comments but not blocking at all.

File: test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy
Comment (dhiaayachi) [inline test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy:7]: How this get into the CI?

File: test/integration/consul-container/libs/cluster/cluster.go
Comment (DanStough) [reply test/integration/consul-container/libs/cluster/cluster.go:152]: Not sure I want to do two passes (ones to get the servers, the other to filter out clients), but I'll refactor the API call into another function and update `Followers`, `Clients`, and `Servers` to use it.

Comment (DanStough) [reply test/integration/consul-container/libs/cluster/cluster.go:152]: Actually, I'm already parsing this `consul-container.go`, so I'll just add a method to the Node interface to expose it.

File: test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy
Comment (DanStough) [reply test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy:7]: The file or the base image? 
* The base image happens in [prior step in the pipeline](https://github.com/hashicorp/consul/blob/4c359dcb696dfbdf234d2b68c756cf1dd137d65b/.circleci/config.yml#L840-L848).
* The file is passed into testcontainers [docker context via go:embed](https://github.com/hashicorp/consul/pull/15084/files#diff-a20cab050471215df956d350bb1352623fa422052f64add57fb0bee080e27818R17-R18), then built dynamically when the test runs. This seemed similar enough to how the bats tests work.

**Comment by DanStough** [comment]
I'm seeing two test failures, but I think these are known issues that are currently being working. Waiting to merge until those are resolved and I can rebase.

**Comment by DanStough** [comment]
>  I'm seeing two test failures, but I think these are known issues that are currently being working. Waiting to merge until those are resolved and I can rebase.

Yes, I confirmed some fixes we have in the pipeline address these issues. Still waiting until they are merged to rebase, but this PR is ready.

File: test/integration/consul-container/libs/node/node.go
Comment (dhiaayachi) [reply test/integration/consul-container/libs/node/node.go]: yes Agent is good!

File: test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy
Comment (dhiaayachi) [reply test/integration/consul-container/libs/service/assets/Dockerfile-consul-envoy:7]: oh I missed the go:embed directive 😓

File: test/integration/consul-container/libs/agent/container.go
Comment (dhiaayachi) [inline test/integration/consul-container/libs/agent/container.go:66]: a nitpick, maybe we can start with the exported functions and leave the non exported at the end

**Review by dhiaayachi — APPROVED** [review]
LGTM!
I had few aesthetics comments but nothing blocking

File: test/integration/consul-container/libs/agent/container.go
Comment (DanStough) [reply test/integration/consul-container/libs/agent/container.go:66]: yep, good call.

File: test/integration/consul-container/libs/agent/container.go
Comment (DanStough) [reply test/integration/consul-container/libs/agent/container.go:74]: I use it in two places:
1. For the `hostname` of the container. The big problem I ran into with `auto-encrypt` was that agents need to specify `retry-join` for it to work, meaning the names must be deterministic, but the container names also need to be random to allow parallel runs. My solution was to use random container names, but given them an `agent-X` hostname that would be unique per cluster. This also works because there is a docker network per-cluster that sandboxes the hostnames. I'd like to see this a bit more automated/intuitive, but I was going to leave that to future work.
1. As you said, for the convenience of debugging. Specifically for servers being rotated, it's hard to tell which containers are "new" vs "old", and which ones new agents should be trying to join.

File: test/integration/consul-container/libs/agent/container.go
Comment (DanStough) [reply test/integration/consul-container/libs/agent/container.go:131]: Similar answer to the question above. We need a predictable hostname for agent `retry-joins`.

File: test/integration/consul-container/test/peering/server_test.go
Comment (DanStough) [reply test/integration/consul-container/test/peering/server_test.go:51]: Good call.

**Pull request merged**
